Enhancement/5692 vulnerability scanner documentation (#29506)
wazuh/wazuh: docs/ref/modules/vulnerability-scanner/test-tools.md

# Test tools

> Below there is a quick overview about the usage of the vulnerability scanner command line tools that serve the purpose of debugging and testing. 

Reviewer: ```suggestion
Below is a quick overview of the usage of the vulnerability scanner command line tools that serve the purpose of debugging and testing. 
```

These tools ease the task of mocking and replicating a real environment where we can validate that new changes do not affect in an undesired way the current capabilities of the module. 

## Compilation 

Command line test tools are intended for development purposes and they are not delivered in the Wazuh manager packages. To use them, it is required to compile the project by sources.

```console
cd wazuh/src
make deps 
make -j$(nproc) TARGET=server
``` 

## Vulnerability Scanner tool 

Location path 

```console
src/build/wazuh_modules/vulnerability_scanner/testtool/scanner/vd_scanner_testtool
```

**Note**: execute the cli with **--help** to display the available options.

### Database creation 

Command

```console
src/build/wazuh_modules/vulnerability_scanner/testtool/scanner/vd_scanner_testtool -c config.json -d
```

Configuration file

```json
{
  "vulnerability-detection": {
    "enabled": "yes",
    "index-status": "no",
    "cti-url": "https://cti.wazuh.com/api/v1/catalog/contexts/vd_1.0.0/consumers/vd_4.8.0"
  },
  "clusterName": "cluster01",
  "clusterEnabled": false
}
```

> - The **-d** flag performs a snapshot download from CTI and process the content to create a local CVE database.

Reviewer: ```suggestion
- The **-d** flag performs a snapshot download from CTI and processes the content to create a local CVE database.
```

- In a real Wazuh Manager installation, **clusterName** is the hostname for a single node deployment, or the cluster name for a cluster setup.
- The outcome of this command is a RocksDB CVE database **queue** in the current directory. 
- The process will take some minutes until the snapshot file is processed.

> **Note:** It's possible to follow the process with the logs printed by the tool. Optional with **-l** the logs can be dumped to a file. 

Reviewer: ```suggestion
**Note:** It's possible to follow the process with the logs printed by the tool. Optionally, with the **-l* argument,* the logs can be dumped to a file. 
```

```console
...
wazuh-modulesd:vulnerability-scanner:databaseFeedManager.hpp:254 processMessage : Processing line: 180001
wazuh-modulesd:vulnerability-scanner:databaseFeedManager.hpp:254 processMessage : Processing line: 181001
wazuh-modulesd:vulnerability-scanner:databaseFeedManager.hpp:254 processMessage : Processing line: 182001
wazuh-modulesd:vulnerability-scanner:databaseFeedManager.hpp:254 processMessage : Processing line: 183001
wazuh-modulesd:vulnerability-scanner:databaseFeedManager.hpp:254 processMessage : Processing line: 184001
wazuh-modulesd:vulnerability-scanner:databaseFeedManager.hpp:254 processMessage : Processing line: 185001
wazuh-modulesd:vulnerability-scanner:databaseFeedManager.hpp:254 processMessage : Processing line: 186001
wazuh-modulesd:vulnerability-scanner:databaseFeedManager.hpp:254 processMessage : Processing line: 187001
...
wazuh-modulesd:vulnerability-scanner:databaseFeedManager.hpp:254 processMessage : Processing line: 292001
wazuh-modulesd:vulnerability-scanner:databaseFeedManager.hpp:369 operator() : Feed update process completed.
Only download content flag is set. Stopping the scanner...
```

> **Note:** Starting the vulnerability scanner tool without **-d** option will update the CVE database with new data not present in the snapshot.

Reviewer: ```suggestion
**Note:** Starting the vulnerability scanner tool without the **-d** option will update the CVE database with new data not present in the snapshot.
```

```console
wazuh-modulesd:vulnerability-scanner:databaseFeedManager.hpp:355 operator() : Initiating update feed process.
wazuh-modulesd:vulnerability-scanner:databaseFeedManager.hpp:188 processMessage : Processing file: queue/vd_updater/tmp/contents/1906695-api_file.json
wazuh-modulesd:vulnerability-scanner:databaseFeedManager.hpp:188 processMessage : Processing file: queue/vd_updater/tmp/contents/1907695-api_file.json
...
wazuh-modulesd:vulnerability-scanner:databaseFeedManager.hpp:369 operator() : Feed update process completed.
```

### Detection 

For details about event format please refer to [Events](events.md)

Command

```console
src/build/wazuh_modules/vulnerability_scanner/testtool/scanner/vd_scanner_testtool -c config.json -u -i os.json,package.json
```

> - That will detect vulnerabilities and fill the indexer databases. 

Reviewer: ```suggestion
- The command will detect vulnerabilities and fill the indexer databases. 
```

> **Note:** It is possible to connect an instance of Wazuh-Indexer and index the result of the vulnerability scan by adding the **Indexer Connector** configuration (more details [Configuration](configuration.md)).

Reviewer: ```suggestion
**Note:** It is possible to connect an instance of the Wazuh-Indexer and index the result of the vulnerability scan by adding the **Indexer Connector** configuration (more details [Configuration](configuration.md)).
```

```console
src/build/wazuh_modules/vulnerability_scanner/testtool/scanner/vd_scanner_testtool -c config.json -t index-template.json -u -i os_event.json,package_event.json
```

Expanded configuration options for indexing

```json
{
  "vulnerability-detection": {
    "enabled": "yes",
    "index-status": "yes",
    "cti-url": "https://cti.wazuh.com/api/v1/catalog/contexts/vd_1.0.0/consumers/vd_4.8.0"
  },
  "indexer": {
    "enabled": "yes",
    "hosts": [
      "https://0.0.0.0:9200"
    ],
    "username": "admin",
    "password": "admin",
    "ssl": {
      "certificate_authorities": [
        "/PATH/root-ca.pem"
      ],
      "certificate": "/PATH/node-1.pem",
      "key": "/PATH/node-1-key.pem"
    },
    "update_mappings_path": ""
  },
  "clusterName": "cluster01",
  "clusterEnabled": false
}
```

**Note:** It is important to modify accordingly the following fields:

  - hosts
  - certificate 
  - key 
  - certificate_authorities 
  - username
  - password

> Also it is required a template to create the mapping in the Wazuh Indexer.

Reviewer: ```suggestion
Also, a template is required to create the mapping in the Wazuh Indexer.
```

indexer-connector:indexerConnector.cpp:606 operator() : Added document for insertion with id: 001_f21aca719022f009d80bbf9224741d79029b31f2_CVE-2024-12243.
indexer-connector:indexerConnector.cpp:606 operator() : Added document for insertion with id: 001_f21aca719022f009d80bbf9224741d79029b31f2_CVE-2024-28835.
indexer-connector:indexerConnector.cpp:606 operator() : Added document for insertion with id: 001_f21aca719022f009d80bbf9224741d79029b31f2_CVE-2024-28834.
indexer-connector:indexerConnector.cpp:606 operator() : Added document for insertion with id: 001_f21aca719022f009d80bbf9224741d79029b31f2_CVE-2024-12243.
indexer-connector:indexerConnector.cpp:606 operator() : Added document for insertion with id: 001_f21aca719022f009d80bbf9224741d79029b31f2_CVE-2024-28835.
indexer-connector:indexerConnector.cpp:606 operator() : Added document for insertion with id: 001_f21aca719022f009d80bbf9224741d79029b31f2_CVE-2024-28834.
indexer-connector:indexerConnector.cpp:606 operator() : Added document for insertion with id: 001_f21aca719022f009d80bbf9224741d79029b31f2_CVE-2024-12243.
```

**Note**: To request the information please refer to the [Opensearch API](api-reference.md)

> **Note**: The **-u** option disables **Content Updater** module, so it won't process any snapshot nor offsets and it will directly process incoming events looking for vulnerabilities.

Reviewer: ```suggestion
**Note**: The **-u** option disables the **Content Updater** module, so it won't process any snapshots or offsets, and it will directly process incoming events looking for vulnerabilities.
```

> **Note**: A package event needs an OS event, if we send a package event without the OS event, the tool may hang even if we later use the right command with OS event. To overcome this, remove the **queue/vd/event** or use the **-b** option explained below.

Reviewer: ```suggestion
**Note**: A package event needs an OS event; if we send a package event without the OS event, the tool may hang even if we later use the right command with the OS event. To overcome this, remove the **queue/vd/event** or use the **-b** option explained below.
```
